Sources/TokamakCore/Shapes/Path.swift

```diff
@@ -0,0 +1,290 @@
+// Copyright 2020 Tokamak contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//  Created by Carson Katri on 06/28/2020.
+//
+
+/// The outline of a 2D shape.
+public struct Path: Equatable, LosslessStringConvertible {
+  public var description: String {
+    """
+    \(storage)
+    \(elements)
+    \(transform)
+    """
+  }
+
+  public enum Storage: Equatable {
+    case empty
+    case rect(CGRect)
+    case ellipse(CGRect)
+    indirect case roundedRect(FixedRoundedRect)
+    indirect case stroked(StrokedPath)
+    indirect case trimmed(TrimmedPath)
+//    case path(PathBox)
+  }
+
+  public enum Element: Equatable {
+    case move(to: CGPoint)
+    case line(to: CGPoint)
+    case quadCurve(to: CGPoint, control: CGPoint)
+    case curve(to: CGPoint, control1: CGPoint, control2: CGPoint)
+    case closeSubpath
+  }
+
+  public var storage: Storage
+  public var elements: [Element] = []
+  public var transform: CGAffineTransform = .identity
+
+  public struct _SubPath: Equatable {
+    let path: Path
+    let transform: CGAffineTransform
+  }
+
+  public var subpaths: [_SubPath] = []
+
+  public init() {
+    storage = .empty
+  }
+
+  init(storage: Storage) {
+    self.storage = storage
+  }
+
+  public init(_ rect: CGRect) {
+    storage = .rect(rect)
+  }
+
+  public init(roundedRect rect: CGRect,
+              cornerSize: CGSize,
+              style: RoundedCornerStyle = .circular) {
+    storage = .roundedRect(FixedRoundedRect(rect: rect,
+                                            cornerSize: cornerSize,
+                                            style: style))
+  }
+
+  public init(roundedRect rect: CGRect,
+              cornerRadius: CGFloat,
+              style: RoundedCornerStyle = .circular) {
+    storage = .roundedRect(FixedRoundedRect(rect: rect,
+                                            cornerSize: CGSize(width: cornerRadius, height: cornerRadius),
+                                            style: style))
+  }
+
+  public init(ellipseIn rect: CGRect) {
+    storage = .ellipse(rect)
+  }
+
+  public init(_ callback: (inout Self) -> ()) {
+    var base = Self()
+    callback(&base)
+    self = base
+  }
+
+  public init?(_ string: String) {
+    // FIXME: Somehow make this from a string?
```

Comment:
Here’s what `Path(roundedRect: CGRect(x: 0, y: 0, width: 50, height: 100), cornerRadius: 5, style: .circular)` looks like:

```
"50 50 m 50 95 l 50 97.7614 47.7614 100 45 100 c 5 100 l 2.23858 100 3.38177e-16 97.7614 0 95 c 5.55112e-15 5 l 5.88929e-15 2.23858 2.23858 -1.69088e-16 5 0 c 45 0 l 47.7614 1.69088e-16 50 2.23858 50 5 c h"
```

and here’s `Path(ellipseIn: CGRect(origin: .zero, size: CGSize(width: 100, height: 100)))`:

```
"100 50 m 100 77.6142 77.6142 100 50 100 c 22.3858 100 0 77.6142 0 50 c 0 22.3858 22.3858 0 50 0 c 77.6142 0 100 22.3858 100 50 c h"
```

not exactly sure what it means 🤷

Reply:
Looks like [path commands](https://developer.mozilla.org/en-US/docs/Web/SVG/Attribute/d)!

Reply:
Sorta but they have all the commands lowercased and it’s a postfix notation

Reply:
Looks to be similar to but different from what you get when you do `String(describing: CGPath(...))` so this can probably just be skipped.

Reply:
You mean that printing the `CGPath` and printing the `Path` are the same? Because I would still need to implement it for `CGPath`...

Reply:
They’re similar but different.

Reply:
Here’s what I’ve got so far (in JS, using PEG.js for parsing because it’s something I’ve used before): https://observablehq.com/d/c4d2be55a2e0b70a

```js
Path = first: Command rest: (s cmd: Command { return cmd })* { return [first, ...rest] }

Command = Move / Line / Curve / Home

Move "move"
  = x: number s y: number s "m"
    { return { command: 'moveto', x, y } }
Line "line"
  = x: number s y: number s "l"
    { return { command: 'lineto', x, y } }
Curve "curve"
  = a: number s b: number s c: number s d: number s e: number s f: number s "c"
    { return { command: 'curveto', args: [a, b, c, d, e, f] } }
Home "home"
  = "h" { return { command: 'closepath' } }

s = " "

number "number"  = "-"? int frac? exp? { return parseFloat(text()) }
exp = [eE] ("+" / "-")? [0-9]+
frac = "." [0-9]+
int = "0" / ([1-9] [0-9]*)
```